File: c++/ValidateBST.cpp
class Solution {
public:
    bool dfs(TreeNode* curNode, long max, long min) {
        if (curNode == NULL) {
            return true;
        }
        
        if ((curNode->val >= max) || (curNode->val <= min)) {
            return false;
        }
        
        return (dfs(curNode->left, curNode->val, min) && dfs(curNode->right, max, curNode->val));
    }

    bool isValidBST(TreeNode* root) {
        return dfs(root, LONG_MAX, LONG_MIN);
    }
};

File: python3/ValidateBST.py
class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        self.isValid = True
        def dfs(curNode, minVal, maxVal):
            if (curNode == None):
                return True
            
            if (curNode.val >= maxVal) or (curNode.val <= minVal):
                return False

            return dfs(curNode.left, minVal, curNode.val) and dfs(curNode.right, curNode.val, maxVal)
        
        return dfs(root, float("-inf"), float("inf"))
